python_modules/dagster/dagster/_daemon/auto_materialize_daemon.py

```diff
@@ -0,0 +1,100 @@
+
+CURSOR_NAME = "ASSET_RECONCILIATION_DAEMON_CURSOR"
+
+
+class AutoMaterializeDaemon(IntervalDaemon):
+    @classmethod
+    def daemon_type(cls) -> str:
+        return "ASSET"
+
+    def run_iteration(
+        self,
+        workspace_process_context: IWorkspaceProcessContext,
+    ) -> DaemonIterator:
+        yield
+
+        instance = workspace_process_context.instance
+        workspace = workspace_process_context.create_request_context()
+        asset_graph = ExternalAssetGraph.from_workspace(workspace)
+
+        target_asset_keys = {
+            target_key
+            for target_key in asset_graph.non_source_asset_keys
+            if asset_graph.get_auto_materialize_policy(target_key) is not None
+        }
+
+        if not target_asset_keys:
+            return
+
+        raw_cursor = instance.run_storage.kvs_get({CURSOR_NAME}).get(CURSOR_NAME)
+        cursor = (
+            AssetReconciliationCursor.from_serialized(raw_cursor, asset_graph)
+            if raw_cursor
+            else AssetReconciliationCursor.empty()
+        )
+
+        run_requests, new_cursor = reconcile(
+            asset_graph=asset_graph,
+            target_asset_keys=target_asset_keys,
+            instance=instance,
+            cursor=cursor,
+            run_tags=None,
+        )
+
+        for run_request in run_requests:
+            asset_keys = check.not_none(run_request.asset_selection)
+            check.invariant(len(asset_keys) > 0)
+
+            repo_handle = asset_graph.get_repository_handle(asset_keys[0])
+
+            # Check that all asset keys are from the same repo
+            for key in asset_keys[1:]:
+                check.invariant(repo_handle == asset_graph.get_repository_handle(key))
+
+            location_name = repo_handle.code_location_origin.location_name
+            repository_name = repo_handle.repository_name
+            job_name = check.not_none(asset_graph.get_implicit_job_name_for_assets(asset_keys))
+
+            code_location = workspace.get_code_location(location_name)
+            external_pipeline = code_location.get_external_pipeline(
+                PipelineSelector(
+                    location_name=location_name,
+                    repository_name=repository_name,
+                    pipeline_name=job_name,
+                    solid_selection=None,
+                    asset_selection=asset_keys,
+                )
+            )
+
+            run = instance.create_run(
+                pipeline_name=external_pipeline.name,
+                run_id=None,
+                run_config={"execution": {"config": {"in_process": {}}}},
+                mode=None,
+                solids_to_execute=None,
+                step_keys_to_execute=None,
+                status=DagsterRunStatus.NOT_STARTED,
+                solid_selection=None,
+                root_run_id=None,
+                parent_run_id=None,
+                tags=run_request.tags,
```

Comment:
We should probably add a tag to indicate that this run comes from this daemon, right?

Reply:
added `dagster/created_by: auto_materialize`. We'll be free to change this later